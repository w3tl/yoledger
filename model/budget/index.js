import Model from '../Model';
import Account from '../accounts';

const schema = {
  type: 'object',
  required: ['account', 'date', 'amount', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    account: {
      type: 'string',
    },
    date: {
      bsonType: 'date',
    },
    amount: {
      type: 'number',
    },
    userId: {
      type: 'string',
    },
  },
};


export default class Budget extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.collName = 'budget';
    this.schema = schema;
  }

  async checkAccount(account) {
    const accountModel = new Account(this.db, this.userId);
    const accountDoc = await accountModel.findByName(account);
    if (!accountDoc) {
      throw new Error('Account not found');
    }
  }

  async addBudget({ account, date, amount }) {
    await this.checkAccount(account);
    const { insertedId } = await super.insertOne({
      date: new Date(date),
      account,
      amount,
    });
    return insertedId;
  }

  async updateBudget(_id, {
    date, account, amount,
  }) {
    const modificator = {};
    if (date) { modificator.date = date; }
    if (account) {
      await this.checkAccount(account);
      modificator.account = account;
    }
    if (amount) { modificator.amount = amount; }

    return super.updateOne({ _id }, { $set: modificator });
  }
}
