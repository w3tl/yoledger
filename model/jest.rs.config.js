module.exports = {
  notify: true,
  transform: {
    '^.+\\.js?$': 'babel-jest',
  },
  globalSetup: './setup.js',
  globalTeardown: './teardown.js',
};
