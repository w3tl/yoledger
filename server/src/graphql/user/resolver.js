export const User = {
  username: ({ _id }) => _id,
};

export const Query = {
  user(root, _, { models: { userModel } }) {
    return userModel.findOne();
  },
};
