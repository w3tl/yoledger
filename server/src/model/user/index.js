import Model from '../Model';
import schema from './schema';
import { createCredentials } from './auth';

export default class User extends Model {
  constructor(db, userId) {
    super(db, userId.toLowerCase());
    this.schema = schema;
    this.collName = 'users';
  }

  async create({ password }) {
    if (await this.findOne({ _id: this.userId })) {
      throw new Error('User exist');
    }

    const credentials = await createCredentials({ password });

    const user = {
      _id: this.userId,
      credentials: {
        current: credentials,
      },
      createdAt: new Date(),
    };

    await super.insertOne(user);
    return this.findOne();
  }

  findOne(...props) {
    return super.findOne({ _id: this.userId }, ...props);
  }
}
