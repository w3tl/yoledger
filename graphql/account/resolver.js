import { Account as Accounts } from '../../model';

export const Account = {
  id: ({ _id }) => _id,
};

export const Query = {
  async accounts(root, { type }, { userId, connection }) {
    const account = new Accounts(connection.db(), userId);

    return account.find({ type }).toArray();
  },
};

export const Mutation = {
  async addAccount(root, { input }, { userId, connection }) {
    const { name, type, balance } = input;
    const account = new Accounts(connection.db(), userId);
    const { insertedId } = await account.create({
      name,
      balance,
      type,
    });

    return {
      account: await account.findOne({ _id: insertedId }),
    };
  },
};
