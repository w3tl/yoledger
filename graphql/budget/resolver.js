import { Budget as Budgets } from '../../model';

export const Budget = {
  async accounts({ periods }, _, { models: { budgetModel } }) {
    const accounts = await budgetModel.aggregate([
      { $match: { date: { $in: periods } } },
      {
        $group: {
          _id: { account: '$account' },
        },
      },
      {
        $lookup: {
          from: 'account',
          localField: '_id.account',
          foreignField: 'name',
          as: 'account',
        },
      },
      {
        $unwind: '$account',
      }, {
        $replaceRoot: { newRoot: '$account' },
      },
    ]).toArray();

    return accounts;
  },
};

export const Query = {
  async budgets(root, { dateStart, count }) {
    if (count > 20) throw new Error('Count too long');
    const periods = Budgets.periods(dateStart, count, {
      type: 'dayOfMonth', options: [10, 25],
    });
    return { periods };
  },
  async budget(root, { date }, { models: { budgetModel } }) {
    return budgetModel.find({ date: new Date(date) }).toArray();
  },
};

export const Mutation = {
  async upsertBudget(root, { input }, { models: { budgetModel } }) {
    const { account, date, amount } = input;
    const id = await budgetModel.upsertBudget({ account, date, amount });
    return {
      success: id,
    };
  },
};

export const Allocation = {
  account({ account }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ name: account, userId });
  },
};
