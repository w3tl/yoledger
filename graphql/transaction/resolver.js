import { GraphQLError } from 'graphql';
import { Transaction as Model } from '../../model';

export const Transaction = {
  id: ({ _id }) => _id,
  date: ({ createdAt }) => createdAt,
  source({ source }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ name: source.name, userId });
  },
  destination({ destination }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ name: destination.name, userId });
  },
};

export const Mutation = {
  async addTransaction(root, { input }, { userId, connection }) {
    const {
      source, destination, amount, date: createdAt,
    } = input;
    const transObj = {
      source,
      destination,
      amount,
      createdAt,
    };

    const trans = new Model(connection.db(), userId);
    let _id;
    await connection.withSession(async (session) => {
      _id = await trans.post(transObj, session);
    });

    const transaction = await trans.findOne({ _id });

    return {
      transaction,
    };
  },
  async deleteTransaction(root, { id }, { userId, connection }) {
    const trans = new Model(connection.db(), userId);
    try {
      await connection.withSession(session => trans.unpost(id, session));
    } catch (err) {
      return {
        success: false,
      };
    }

    return {
      success: true,
    };
  },
  async updateTransaction(root, { id, input }, { userId, connection }) {
    const trans = new Model(connection.db(), userId);
    const {
      amount, source, destination, date: createdAt,
    } = input;
    const updates = {
      source,
      destination,
      amount,
      createdAt,
    };
    try {
      let updatedId;
      await connection.withSession(async (session) => {
        updatedId = await trans.update(id, updates, session);
      });
      const transaction = await trans.findOne({ _id: updatedId });
      return { transaction };
    } catch (err) {
      throw new GraphQLError(err.message);
    }
  },
};

export const Query = {
  /* eslint-disable-next-line object-curly-newline */
  async transactions(_, { dateStart, dateEnd, page, itemsPerPage }, { userId, connection }) {
    const trans = new Model(connection.db(), userId);
    const query = {};
    if (dateEnd) {
      query.$and = [
        { createdAt: { $gte: dateStart } },
        { createdAt: { $lte: dateEnd } },
      ];
    } else {
      query.createdAt = { $gte: dateStart };
    }

    return trans.find(query)
      .skip(page * itemsPerPage)
      .limit(itemsPerPage)
      .toArray();
  },
};
