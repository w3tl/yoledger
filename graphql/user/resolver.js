import { User as Model } from '../../model';

export const User = {
  username: ({ _id }) => _id,
};

export const Query = {
  user(root, _, { userId, connection }) {
    const user = new Model(connection.db(), userId);
    return user.findOne();
  },
};
