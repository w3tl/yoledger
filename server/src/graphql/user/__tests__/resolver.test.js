jest.mock('mongodb');

import { Query, User } from '../resolver';
import { User as UserModel } from '../../../model';

const userId = 'admin';

describe('user resolver', () => {
  const { connection } = require('mongodb');
  const models = { userModel: new UserModel(connection.db(), userId) };
  test('Query', async () => {
    const result = await Query.user(null, {}, { models });
    expect(result).toMatchSnapshot();
  });

  test('User', () => {
    expect(User.username({ _id: userId })).toBe(userId);
  });
});
