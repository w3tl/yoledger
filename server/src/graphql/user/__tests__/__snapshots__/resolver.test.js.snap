// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user resolver Query 1`] = `
Object {
  "_id": "admin",
}
`;
